### ISSUE TITLE
padding issue for CausalConv1d

### ISSUE THREAD

**Issue by YoungloLee** [lucidrains/audiolm-pytorch#166]
https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/soundstream.py#L303-L314
In CausalConv1d class, the number of padding should be (dilation * (kernel_size - 1) + 1 - stride), not just (dilation * (kernel_size - 1)).
Without this correction, the soundstream model does not converge.
Hope this helps you.

`class CausalConv1d(nn.Module):
    def __init__(self, chan_in, chan_out, kernel_size, **kwargs):
        super().__init__()
        kernel_size = kernel_size
        dilation = kwargs.get('dilation', 1)
        stride = kwargs.get('stride', 1)
        self.causal_padding = dilation * (kernel_size - 1) + 1 - stride
        self.conv = nn.Conv1d(chan_in, chan_out, kernel_size, **kwargs)

    def forward(self, x):
        x = F.pad(x, (self.causal_padding, 0), mode='reflect')
        return self.conv(x)`

**Comment by amitaie** [comment]
Hey, can you share your training graphs\results with your correction and without?

**Comment by lucidrains** [comment]
@YoungloLee thank you dearly for this! i believe you are correct and this is a huge misstep on my part :pray:

**Comment by amitaie** [comment]
> Also, i just set the padding mode to be 'reflect' following FAIR's EnCodec implementation

maybe i'm totally worng but doesn't reflect padding mode make it not streaming?

**Comment by YoungloLee** [comment]
> > Also, i just set the padding mode to be 'reflect' following FAIR's EnCodec implementation
> 
> maybe i'm totally worng but doesn't reflect padding mode make it not streaming?

I think it does not matter for streaming.

**Comment by Dannynis** [comment]
Hey, doesnt it affects the CausalTransposedConv1d as well if so ?

**Comment by ksasso1028** [comment]
shouldnt this be
```
class CausalConv1d(nn.Module):
    def __init__(self, chan_in, chan_out, kernel_size, **kwargs):
        super().__init__()
        kernel_size = kernel_size
        dilation = kwargs.get('dilation', 1)
        stride = kwargs.get('stride', 1)
        self.causal_padding = (dilation * (kernel_size - 1))  // stride
        self.conv = nn.Conv1d(chan_in, chan_out, kernel_size, **kwargs)

    def forward(self, x):
        x = F.pad(x, (self.causal_padding, 0), mode='reflect')
        return self.conv(x)
```
since you need to adjust your padding by a factor of stride, not a offset
